data/lightdm-set-defaults.1
.TH LIGHTDM-SET-DEFAULTS 1 "7 November 2013"
.SH NAME
lightdm-set-defaults \- sets configuration options for lightdm
.SH SYNOPSIS
.B lightdm-set-defaults
[
.I OPTION
]
.SH DESCRIPTION
.B lightdm-set-defaults
sets configuration options for lightdm.  Requires root permissions to run.
.SH OPTIONS
.TP
.B \-h, \-\-help
Show help options
.TP
.B \-k, \-\-keep\-old
Only update if no default already set
.TP
.B \-d, \-\-debug
Enable debugging messages
.TP
.B \-r, \-\-remove
Remove default value if it's the current one
.TP
.B \-s, \-\-session=SESSION_NAME
Set the default session to SESSION_NAME
.TP
.B \-g, \-\-greeter=GREETER_NAME
Set the default greeter to GREETER_NAME
.TP
.B \-a, \-\-autologin=USERNAME
Set the autologin user to USERNAME
.TP
.B \-i, \-\-hide\-users=<true|false>
Set greeter-hide-users to true or false.
.TP
.B \-m, \-\-show\-manual\-login=<true|false>
Set show-manual-login to true or false.
.TP
.B \-R, \-\-show\-remote\-login=<true|false>
Set show-manual-login to true or false.
.TP
.B \-l, \-\-allow\-guest=<true|false>
Set allow-guest to true or false.
.SH FILES
.TP
.B /etc/lightdm/lightdm.conf
Configuration
.SH SEE ALSO
.BR lightdm (1),
.BR dm-tool (1)


NEWS
Overview of changes in lightdm 1.8.5

    * Don't pass system user accounts from AccountsService to greeters.
    * Fix crash if switching to greeter and it isn't installed.

Overview of changes in lightdm 1.8.4

    * Implement missing guest-wrapper functionality and enable it for Ubuntu.
    * Update AppArmor scripts to work in Ubuntu 13.10.

Overview of changes in lightdm 1.8.3

    * Correctly set $XDG_SESSION_CLASS for greeters. This was regressed in 1.7.5
      for ConsoleKit and was never passed to logind. logind/ConsoleKit treat
      greeter sessions without this set as user sessions. This causes greeters
      to show the lightdm user able to be logged in with.
    * Set $USER when running the session-setup-script. This is a regression from
      1.7.5.
    * Fix notification of sessions being logged out. This is a regression from
      1.7.5 and caused greeters to show sessions logged in after they had been
      logged out.

Overview of changes in lightdm 1.8.2

    * Fix crash when starting with existing X servers. This was introduced in
      rev 1651 (lightdm 1.7.0).
  
Overview of changes in lightdm 1.8.1

    * Fix crash where Process objects are accessed after unref

Overview of changes in lightdm 1.8.0

    * Add regression test for corrupt X authority files.

Overview of changes in lightdm 1.7.18

    * Set session environment variables for guest sessions (1.7 regression).
    * Don't fail writing X authority if reading it had an error.
    * Update environment variables that we pass to Mir.

Overview of changes in lightdm 1.7.17

    * surfaceflinger: Set XDG_VTNR=0 if VTs are not available
    * Allow compiling of liblightdm-qt without liblightdm-gobject
    * Add missing documentation for xremote seat options.

Overview of changes in lightdm 1.7.16

    * Fix ConsoleKit support broken in 1.7.5
    * Fix --test-mode
    * Add support for running Surfaceflinger sessions

Overview of changes in lightdm 1.7.15

    * Set XDG_VTNR=1 as a fallback if using SeatUnity without a functioning
      compositor or working VT switching, so that logind will recognize the VT
      as active.

Overview of changes in lightdm 1.7.14

    * Correctly set permissions on Xauthority file.

Overview of changes in lightdm 1.7.13

    * Correctly set $XDG_CURRENT_DESKTOP for non-autologin sessions
    * Fix overallocation of array for strings from greeter.
    * Fix truncation writing card32 in XDMCP server.

Overview of changes in lightdm 1.7.12

    * Add xdg-seat config setting
    * Notify Unity System Compositor of the session being authenticated

Overview of changes in lightdm 1.7.11

    * Fix crash when greeter quits due to read watch not being removed
    * Correctly setup Unity System Compositor environment
    * Improve log messages

Overview of changes in lightdm 1.7.10

    * Fix session locking broken in 1.7.5
    * Load lightdm.conf after lightdm.conf.d/*.conf
    * Also support loading config from /usr/share
    * When switching sessions show a greeter if authentication required
    * Set $XDG_CURRENT_DESKTOP if specified in the xsession file
    * Change logging prefixes to make it easier to troubleshoot multiseat setups
    * Bring Ubuntu packaging in-tree

Overview of changes in lightdm 1.7.9

    * Correctly set XDG_VTNR for unity sessions that are not autologin.

Overview of changes in lightdm 1.7.8

    * Add support for Mir sessions and greeters.
    * Set XDG_VTNR for unity sessions.
    * Fix desktop-session-start upstart signal not being emitted since 1.7.5.
    * Fix greeter log broken in 1.7.5.

Overview of changes in lightdm 1.7.7

    * Fix uninitialized pointer introduced in 1.7.3
    * Enable compile warnings and fix code generating warnings

Overview of changes in lightdm 1.7.6

    * Restore greeter hints that were regressed in 1.7.5.
    * Don't run greeters through session wrapper - regression in 1.7.5

Overview of changes in lightdm 1.7.5

    * Quit Plymouth correctly when using the unity seat type
    * Release the VT when the system compositor fails to start
    * Load sessions and greeters from /usr/share/lightdm/sessions and
      /usr/share/lightdm/greeters. The existing directories are checked
      if the sessions are not in these directories.
    * Refactor the Display class so that it merges with the Seat class
    * Support running the greeter and session in different display servers
      instead of re-using the same one during a login.
    * Add more regression tests
    * Documentation fixes

Overview of changes in lightdm 1.7.4

    * Set XDG_SEAT and XDG_VTNR environment variables.
    * Add initial support for Unity (i.e. Mir based) seats.
    * Add a greeter wrapper option.

Overview of changes in lightdm 1.7.3

    * Load configuration from /etc/lightdm/lightdm.conf.d.
    * Fix compile warnings
    * Fix tests not running from install directory inside checkout.

Overview of changes in lightdm 1.7.2

    * Fix incorrectly distributed guest-session apparmor data

Overview of changes in lightdm 1.7.1

    * Fix .pc file for liblightdm-qt5-3
    * Add a new option "autologin-in-background" which lets an autologin happen
      in a second display while still showing the greeter.
    * Stop if fail to create default seat
    * Add option to set seat type in lightdm-set-defaults
    * Stop using g_file_set_contents - it can leave intermediate files around
    * Make tests work without installing them
    * Fix distcheck

Overview of changes in lightdm 1.7.0

    * Use logind instead of ConsoleKit if it is available
    * Use Q_SLOTS and Q_SIGNALS instead of slots and signals.
    * Ignore stale X server locks
    * Pass through system locale or set locale from AccountsService/.dmrc
    * Fix bug where seat failure before D-Bus acquired would not stop daemon

Overview of changes in lightdm 1.6.0

    * Allow VNC command to be specified in lightdm.conf
    * Register enums with QObject meta type system.

Overview of changes in lightdm 1.5.3


utils/lightdm-set-defaults.c
/*
 * Copyright (C) 2011 Didier Roche.
 * Author: Didier Roche <didrocks@ubuntu.com>
 * 
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version. See http://www.gnu.org/copyleft/gpl.html the full text of the
 * license.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <unistd.h>

#include <glib.h>
#include <glib/gi18n.h>

#define SEATDEFAULT_KEY_GROUP "SeatDefaults"
#define TYPE_KEY_NAME  "type"
#define SESSION_KEY_NAME  "user-session"
#define GREETER_KEY_NAME  "greeter-session"
#define AUTOLOGIN_KEY_NAME  "autologin-user"
#define HIDE_USERS_KEY_NAME  "greeter-hide-users"
#define MANUAL_LOGIN_KEY_NAME  "greeter-show-manual-login"
#define REMOTE_LOGIN_KEY_NAME  "greeter-show-remote-login"
#define ALLOW_GUEST_KEY_NAME  "allow-guest"

#define IS_STRING_EMPTY(x) ((x)==NULL||(x)[0]=='\0')

static gboolean debug = FALSE;
static gboolean keep_old = FALSE;
static gboolean remove = FALSE;
static gboolean hide_users = FALSE;
static gboolean show_manual_login = FALSE;
static gboolean show_remote_login = FALSE;
static gboolean allow_guest = FALSE;

static char    *type = NULL;
static char    *session = NULL;
static char    *greeter = NULL;
static char    *autologin = NULL;
static char    *str_hide_users = NULL;
static char    *str_show_manual_login = NULL;
static char    *str_show_remote_login = NULL;
static char    *str_allow_guest = NULL;

static GOptionEntry entries[] =
{
  { "debug",    'd', 0, G_OPTION_ARG_NONE, &debug, N_("Enable debugging"), NULL },
  { "keep-old", 'k', 0, G_OPTION_ARG_NONE, &keep_old, N_("Only update if no default already set"), NULL },
  { "remove",   'r', 0, G_OPTION_ARG_NONE, &remove, N_("Remove default value if it's the current one"), NULL },
  { "type",     't', 0, G_OPTION_ARG_STRING, &type, N_("Set default seat type"), NULL },
  { "session",  's', 0, G_OPTION_ARG_STRING, &session, N_("Set default session"), NULL },
  { "greeter",  'g', 0, G_OPTION_ARG_STRING, &greeter, N_("Set default greeter"), NULL },
  { "autologin",'a', 0, G_OPTION_ARG_STRING, &autologin, N_("Set autologin user"), NULL },
  { "hide-users",'i', 0, G_OPTION_ARG_STRING, &str_hide_users, N_("Set greeter-hide-users to true or false"), NULL },
  { "show-manual-login",'m', 0, G_OPTION_ARG_STRING, &str_show_manual_login, N_("Set show-manual-login to true or false"), NULL },
  { "show-remote-login",'R', 0, G_OPTION_ARG_STRING, &str_show_remote_login, N_("Set show-remote-login to true or false"), NULL },
  { "allow-guest",'l', 0, G_OPTION_ARG_STRING, &str_allow_guest, N_("Set allow-guest to true or false"), NULL },
  { NULL }
};

void
show_nothing(const gchar   *log_domain,
             GLogLevelFlags log_level,
             const gchar   *message,
             gpointer       unused_data) {};

int
update_boolean(const gboolean new_value,
              gboolean     keep_old,
              const gchar *key_group,
              const gchar *key_name,
              GKeyFile    *keyfile)
{
    if (keep_old)
        g_debug ("keep-old mode: keep previous default value");
    else {
        g_debug ("Update to %d for %s", new_value, key_name);
        g_key_file_set_boolean (keyfile, key_group, key_name, new_value);
    }
    return(0);
}

int
update_string(const gchar *default_value,
              const gchar *new_value,
              gboolean     keep_old,
              gboolean     remove,
              const gchar *key_group,
              const gchar *key_name,
              GKeyFile    *keyfile)
{
    if (!(default_value) || (strlen(default_value) < 1)) {
        g_debug ("No existing valid value for %s. Set to %s", key_name, new_value);
        g_key_file_set_string (keyfile, key_group, key_name, new_value);
    }
    else {
        if (remove) {
            if (g_strcmp0 (default_value, new_value) == 0) {
                g_debug ("Remove %s as default value for %s", default_value, key_name);
                g_key_file_set_string (keyfile, key_group, key_name, "");
                return(0);
            }
            g_debug ("Can't remove: %s is not the default value for %s", default_value, key_name);
            return(4);
        }
        else {
            g_debug ("Found existing default value(%s) for %s", default_value, key_name);
            if (keep_old)
                g_debug ("keep-old mode: keep previous default value");
            else {
                g_debug ("Update to %s for %s", new_value, key_name);
                g_key_file_set_string (keyfile, key_group, key_name, new_value);
            }
        }
    }
    return(0);
}

int 
str_to_bool(const gchar *str, gboolean *bool_out)
{
    if (IS_STRING_EMPTY(str)) {
        return -1;
    }
    else if (strncasecmp(str, "true", 4)==0) {
        *bool_out = TRUE;
        return 0;
    }
    else if (strncasecmp(str, "false", 5)==0) {
        *bool_out = FALSE;
        return 0;
    }
    else {
        return -2;
    }
}

int 
main (int argc, char *argv[])
{
    GOptionContext *context = NULL;
    GError         *error = NULL;

    GKeyFile       *keyfile;
    GKeyFileFlags   flags;
    gchar          *s_data;
    gsize           size;
    const gchar    *gdm_conf_file = CONFIG_DIR "/lightdm.conf";

    gchar          *default_type = NULL;
    gchar          *default_session = NULL;
    gchar          *default_greeter = NULL;
    gchar          *default_autologin = NULL;
    gint            return_code = 0;

    bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
    bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
    textdomain (GETTEXT_PACKAGE);

#if !defined(GLIB_VERSION_2_36)
    g_type_init ();
#endif

    context = g_option_context_new (N_("- set lightdm default values"));
    g_option_context_add_main_entries (context, entries, NULL);
    if (!g_option_context_parse (context, &argc, &argv, &error)) {
        g_printerr (N_("option parsing failed: %s\n"), error->message);
        g_option_context_free (context);
        g_error_free (error);
        return 1;
    }
    if (IS_STRING_EMPTY (type) && IS_STRING_EMPTY (session) && IS_STRING_EMPTY (greeter) && IS_STRING_EMPTY (autologin) && IS_STRING_EMPTY(str_hide_users) && IS_STRING_EMPTY(str_show_manual_login) && IS_STRING_EMPTY(str_show_remote_login) && IS_STRING_EMPTY(str_allow_guest)) {
        g_printerr (N_("Wrong usage of the command\n%s"), g_option_context_get_help (context, FALSE, NULL));
        g_option_context_free (context);
        return 1;
    }
    if (context)
        g_option_context_free (context); 
    if (!debug)
        g_log_set_handler (NULL, G_LOG_LEVEL_DEBUG, show_nothing, NULL);

    keyfile = g_key_file_new ();
    flags = G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS;
    if (!(g_key_file_load_from_file (keyfile, gdm_conf_file, flags, &error))) {
            g_debug ("File doesn't seem to exist or can't be read: create one (%s)", error->message);
            g_error_free (error);
            error = NULL;
    }

    // try to get the right keys
    default_type = g_key_file_get_string (keyfile, SEATDEFAULT_KEY_GROUP, TYPE_KEY_NAME, NULL);
    default_session = g_key_file_get_string (keyfile, SEATDEFAULT_KEY_GROUP, SESSION_KEY_NAME, NULL);
    default_greeter = g_key_file_get_string (keyfile, SEATDEFAULT_KEY_GROUP, GREETER_KEY_NAME, NULL);
    default_autologin = g_key_file_get_string (keyfile, SEATDEFAULT_KEY_GROUP, AUTOLOGIN_KEY_NAME, NULL);

    if (!(IS_STRING_EMPTY (type)))
        return_code = update_string (default_type, type, keep_old, remove, SEATDEFAULT_KEY_GROUP, TYPE_KEY_NAME, keyfile);
    if (!(IS_STRING_EMPTY (session)))
        return_code = update_string (default_session, session, keep_old, remove, SEATDEFAULT_KEY_GROUP, SESSION_KEY_NAME, keyfile);
    if (!(IS_STRING_EMPTY (greeter)) && (return_code == 0))
        return_code = update_string (default_greeter, greeter, keep_old, remove, SEATDEFAULT_KEY_GROUP, GREETER_KEY_NAME, keyfile);
    if (!(IS_STRING_EMPTY (autologin)) && (return_code == 0))
        return_code = update_string (default_autologin, autologin, keep_old, remove, SEATDEFAULT_KEY_GROUP, AUTOLOGIN_KEY_NAME, keyfile);
    if (!(IS_STRING_EMPTY(str_hide_users)) && (return_code == 0)) {
        if (str_to_bool(str_hide_users, &hide_users) == 0) {
            return_code = update_boolean (hide_users, keep_old, SEATDEFAULT_KEY_GROUP, HIDE_USERS_KEY_NAME, keyfile);
        }
        else {
            g_printerr (N_("true and false are the only valid choices for hide-users\n"));
            return 1;
        }
    }
    if (!(IS_STRING_EMPTY(str_allow_guest)) && (return_code == 0)) {
        if (str_to_bool(str_allow_guest, &allow_guest) == 0) {
            return_code = update_boolean (allow_guest, keep_old, SEATDEFAULT_KEY_GROUP, ALLOW_GUEST_KEY_NAME, keyfile);
        }
        else {
            g_printerr (N_("true and false are the only valid choices for allow-guest\n"));
            return 1;
        }
    }
    if (!(IS_STRING_EMPTY(str_show_manual_login)) && (return_code == 0)) {
        if (str_to_bool(str_show_manual_login, &show_manual_login) == 0) {
            return_code = update_boolean (show_manual_login, keep_old, SEATDEFAULT_KEY_GROUP, MANUAL_LOGIN_KEY_NAME, keyfile);
        }
        else {
            g_printerr (N_("true and false are the only valid choices for show-manual-login\n"));
            return 1;
        }
    }
    if (!(IS_STRING_EMPTY(str_show_remote_login)) && (return_code == 0)) {
        if (str_to_bool(str_show_remote_login, &show_remote_login) == 0) {
            return_code = update_boolean (show_remote_login, keep_old, SEATDEFAULT_KEY_GROUP, REMOTE_LOGIN_KEY_NAME, keyfile);
        }
        else {
            g_printerr (N_("true and false are the only valid choices for show-remote-login\n"));
            return 1;
        }
    }

    if (return_code == 0) {
        s_data = g_key_file_to_data (keyfile, &size, &error);
        if (!s_data) {
            g_debug ("Can't convert data to string: %s", error->message);
            g_error_free (error);
            return_code = 1;
        }
        else {
            if(!g_file_set_contents (gdm_conf_file, s_data, size, &error)) {
                g_printerr ("Can't update: %s\n", error->message);
                g_error_free (error);
                return_code = 1;
            }
            g_free (s_data);
         }
    }

    g_key_file_free (keyfile);

    if (default_type)
        g_free (default_type);
    if (default_session)
        g_free (default_session);
    if (default_greeter)
        g_free (default_greeter);
    if (default_autologin)
        g_free (default_autologin);

    return return_code;

}
